src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -447,43 +622,683 @@ impl<'tcx> Constructor<'tcx> {
+                        // behave the same (i.e. will be matched only by wildcards), we return a
+                        // meta-constructor that contains all of them at once.
+                        smallvec![MissingConstructors(missing_ctors)]
+                    }
+                } else {
+                    // Here we know there are no missing constructors, so we have to try all
+                    // existing constructors one-by-one.
+                    let (all_ctors, _) = missing_ctors.into_inner();
+                    // Recursively split newly generated list of constructors. This list must not
+                    // contain any wildcards so we don't recurse infinitely.
+                    all_ctors
+                        .into_iter()
+                        .flat_map(|ctor| ctor.split_meta_constructor(cx, ty, head_ctors))
+                        .collect()
+                }
+            }
+            MissingConstructors(_) => bug!("shouldn't try to split constructor {:?}", self),
+        }
+    }
+
+    /// Returns a collection of constructors that spans the constructors covered by `self`,
+    /// subtracted by the constructors covered by `head_ctors`: i.e., `self \ head_ctors` (in set
+    /// notation).
+    fn subtract_meta_constructor(
+        self,
+        tcx: TyCtxt<'tcx>,
+        param_env: ty::ParamEnv<'tcx>,
+        used_ctors: &Vec<Constructor<'tcx>>,
+    ) -> SmallVec<[Constructor<'tcx>; 1]> {
+        debug!("subtract_meta_constructor {:?}", self);
+        // The input must not contain a wildcard
+        assert!(!used_ctors.iter().any(|c| c.is_wildcard()));
+
+        match self {
+            // Those constructors can't intersect with a non-wildcard meta-constructor, so we're
+            // fine just comparing for equality.
+            Single | Variant(_) => {
+                if used_ctors.iter().any(|c| c == &self) {
+                    smallvec![]
+                } else {
+                    smallvec![self]
+                }
+            }
+            FixedLenSlice(self_len) => {
+                let overlaps = |c: &Constructor<'_>| match c {
+                    FixedLenSlice(other_len) => *other_len == self_len,
+                    VarLenSlice(prefix, suffix) => prefix + suffix <= self_len,
+                    _ => false,
+                };
+                if used_ctors.iter().any(overlaps) { smallvec![] } else { smallvec![self] }
+            }
+            VarLenSlice(self_prefix, self_suffix) => {
+                // Assume we have the following match:
+                // ```
+                // match slice {
+                //     [0] => {}
+                //     [_, _, _] => {}
+                //     [1, 2, 3, 4, 5, 6, ..] => {}
+                //     [_, _, _, _, _, _, _, _] => {}
+                //     [0, ..] => {}
+                // }
+                // ```
+                // We want to know which constructors are matched by the last pattern, but are not
```

Comment:
This algorithm feels a bit over-engineered, given that it is OK to take time proportional to `sizeof(FixedLenSlice)` - we already have to eat that much wildcard patterns.

Maybe it would be better to merge the 2 places we are doing the "range union and intersection" stuff?

Reply:
e.g.,
```rust

let our_min_len = self_prefix + self_suffix;
let fixed_lengths: FxHashSet<u64> = used_ctors
    .iter()
    .filter_map(|c: &Constructor<'tcx>| match c {
        FixedLenSlice(n) => n,
        _ => None,
    })
    .collect(); 
let min_var_len: Option<_> = used_ctors
    .iter()
    .filter_map(|c: &Constructor<'tcx>| match c {
        VarLenSlice(prefix, suffix) => Some(prefix + suffix),
        _ => None,
    })
    .min();
    
if let Some(min_var_len) = min_var_len {
    (0..min_var_len)
        .filter(|l| !fixed_lengths.contains(l))
        .map(FixedLenSlice).collect()
} else {
    let min_free_len = Some(self_suffix)
        .chain(fixed_lengths.iter().flat_map(|x| x+1))
        .max().unwrap();
    // Yes I know this misses `FixedLenSlice`.
    (0..self_prefix+min_free_suffix_len)
        .filter(|l| !fixed_lengths.contains(l))
        .map(FixedLenSlice)
        .chain(VarLenSlice(self_prefix, max(min_free_len, our_min_len) - self_prefix))
        .collect()
}
```

Reply:
If you are sufficiently convinced the current code is well-debugged, I'm OK with it merging without it being replaced.

Reply:
I agree that it is over-engineered. I like your suggestion, it looks easier to understand